### PR TITLE
[ADR-13]: Add ValidityInterval.

### PR DESCRIPTION
**Pull request by ak3n**
Implements https://github.com/input-output-hk/plutus-apps/blob/main/doc/adr/0013-tx-validity-time-range-fix.rst

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [ ] Formatting, PNG optimization, etc. are updated
    - [x] Important changes are reflected in changelog.d of the affected packages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reference the ADR in the PR and reference the PR in the ADR (if revelant)
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-ledger-constraints/src/Ledger/Constraints/ValidityInterval.hs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+-- | @from a@ is an 'ValidityInterval' that includes all values that are
+--  greater than or equal to @a@. In math. notation: [a,+∞]
+from :: a -> ValidityInterval a
+from s = ValidityInterval (Just s) Nothing
+
+{-# INLINABLE to #-}
+-- | @to a@ is an 'ValidityInterval' that includes all values that are
+--  smaller than to @a@. In math. notation: [-∞,a)
+to :: a -> ValidityInterval a
+to s = ValidityInterval Nothing (Just s)
```
Comment (berewt) [inline plutus-ledger-constraints/src/Ledger/Constraints/ValidityInterval.hs:50]: The behaviour is aligned with the documentation but I wonder whether the given upper bound should be included in the interval or not: when we count _to_ a given value, we usually include this value.

File: plutus-use-cases/src/Plutus/Contracts.hs [hunk 2]
```diff
@@ -40,34 +40,3 @@ These two steps depend on each other because the transaction fee is a
 function of the size of the transaction including its
 inputs.
 -}
-
-{- Note [Validity Interval's upper bound]
```
Comment (berewt) [inline plutus-use-cases/src/Plutus/Contracts.hs:44]: Do we want to remove this note before the cardano-ledger bug is resolve? ADR-13 gives a clearer semantic for validity range but it doesn't invalidate this comment, does it?

**Review by berewt — APPROVED** [review]
Minor remarks and changes.

File: plutus-ledger-constraints/src/Ledger/Constraints/ValidityInterval.hs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+toLowerBound _        = LowerBound NegInf True
+
+{-# INLINABLE toUpperBound #-}
+toUpperBound :: Maybe a -> UpperBound a
+toUpperBound (Just v) = UpperBound (Finite v) False
+toUpperBound _        = UpperBound PosInf True
+
+{-# INLINABLE toPlutusInterval #-}
+toPlutusInterval :: ValidityInterval a -> Interval a
+toPlutusInterval (ValidityInterval from' to') = Interval (toLowerBound from') (toUpperBound to')
```
[the remark this answers is not in the document]
Comment (berewt) [reply plutus-ledger-constraints/src/Ledger/Constraints/ValidityInterval.hs:78]: a round-trip property test for `fromPlutusInterval . toPlutusInterval` (and the opposite direction) maybe?

File: plutus-contract/test/Spec/Contract/Tx/Constraints/MustSatisfyAnyOf.hs [hunk 4]
```diff
@@ -338,7 +338,7 @@ buildConstraints :: ConstraintParams -> [Cons.TxConstraints () ()]
 buildConstraints cps = do
     let maybeConstraints :: [Maybe (Cons.TxConstraints () ())] = [
             P.maybe Nothing (\cp ->
-                Just $ Cons.mustValidateIn $ to (timeTo cp)) (mustValidateIn cps),
+                Just $ Cons.mustValidateInTimeRange $ Interval.to (P.succ $ timeTo cp)) (mustValidateInTimeRange cps),
```
Comment (sjoerdvisscher) [inline plutus-contract/test/Spec/Contract/Tx/Constraints/MustSatisfyAnyOf.hs:341]: Maybe add a note why you need `succ` here?

**Review by sjoerdvisscher — APPROVED** [review]

**Pull request merged**